### PR TITLE
[OTE][Release][XAI] Detection fix two stage bbox_head error

### PR DESCRIPTION
**Pull request by dongkwan-kim01**
This is the fix of the error issued in [issue](https://github.com/openvinotoolkit/training_extensions/pull/1402#issuecomment-1336631895), which fixes the `simple_test` function in the class`SAMDetectorMixin`.
And resolving the tiling detection length Error regarding the output size assertion.

### CODE REVIEW

**Comment by sungmanc** [comment]
run ote-test

**Comment by sungmanc** [comment]
Check the results at first: https://ci-ote.iotg.sclab.intel.com/job/ote/job/pr-ote-test/809/

**Comment by dongkwan-kim01** [comment]
run ote-test

File: external/model-preparation-algorithm/mpa_tasks/apis/detection/task.py [hunk 1]
```diff
@@ -138,9 +138,9 @@ def _infer_detector(
         output = results["outputs"]
         metric = output["metric"]
         predictions = output["detections"]
-        assert len(output["detections"]) == len(output["feature_vectors"]) == len(output["saliency_maps"]), (
+        assert len(output["detections"]) == len(output["feature_vectors"]), (
             "Number of elements should be the same, however, number of outputs are "
-            f"{len(output['detections'])}, {len(output['feature_vectors'])}, and {len(output['saliency_maps'])}"
+            f"{len(output['detections'])}, {len(output['feature_vectors'])})"
         )
```
Comment (negvet) [inline external/model-preparation-algorithm/mpa_tasks/apis/detection/task.py:144]: Does it cause test failure?

Yes, total number of saliency maps is `len(dataset) * N` where N is the number of classes. But single saliency map is actually a 3-dimentional tensor [N, H, W], where N is the number of classes. So basically `len(output["feature_vectors"]) == len(output["saliency_maps"])` actually should be `True`.

**Comment by sungmanc** [comment]
@dongkwan-kim01, Is there any difference between this PR and [PR#1408](https://github.com/openvinotoolkit/training_extensions/pull/1408)?

**Comment by negvet** [comment]
Are you sure that it fails because of the number of saliency maps?

Basically saliency map tensor created for every dataset point. It has dimensions [N, H, W]. Where N is the total number of classes for torch task or number of predicted classes for the OV task.

But the number of saliency maps should be equal to `len(dataset)` in any case I believe

**Comment by sungmanc** [comment]
> Are you sure that it fails because of the number of saliency maps?
> 
> Basically saliency map tensor created for every dataset point. It has dimensions [N, H, W]. Where N is the total number of classes for torch task or number of predicted classes for the OV task.
> 
> But the number of saliency maps should be equal to `len(dataset)` in any case I believe

Then, could you check this FAIL case?  (http://validationreports.sclab.intel.com:8006/reports/test_result/638e0a31fd6d37c8e261ba3a)

The error log was as below:
image


The same errors were occurred in detection + tiling cases. @eugene123tw do you have any idea for this issue?
image

**Comment by sungmanc** [comment]
Test results: http://validationreports.sclab.intel.com:8006/reports/build_number_report?test_session_build_number=pr-ote-test-810&environment=iotg-dev-workstation-08

Most FAIL cases were solved. Although there were still two FAIL cases, it was not related to this PR. How about merging MPA first and this PR to OTE? @harimkang @dongkwan-kim01 @goodsong81 @negvet

**Review by harimkang — DISMISSED** [review]
Checking the test results, one is a score drop in cls (train vs export) problem, and one is OOM. I think the existing issue appears to have been resolved.

I have confirmed [this](https://ci-ote.iotg.sclab.intel.com/job/ote/job/pr-ote-test/810/)

**Review by goodsong81 — DISMISSED** [review]
LGTM. I've approved MPA PRs.
@sungmanc Could you merge MPA PRs?
@dongkwan-kim01 Could you update MPA commits after merge?

**Comment by dongkwan-kim01** [comment]
I've talked with Evgeny, but in the tile detection case, it seems like many saliency maps are exported for each tile.
So exception handle which `saliency_maps = [saliency_maps[i] for i in range(dataset.num_samples)]`, existed in the previous `develop`, can avoid the assertion error but maybe not make the appropriate saliency map(maybe saliency maps of first few tiles).
Maybe saliency maps merging function or getting the saliency map from the original image(before tiling) would be needed.

> LGTM. I've approved MPA PRs. @sungmanc Could you merge MPA PRs? @dongkwan-kim01 Could you update MPA commits after merge?

Yes, I'll update after the merge but I think we need some fixes in the saliency map for tiling case.. Maybe @negvet investigating and make another PR for it.

**Comment by negvet** [comment]
run ote-test

**Comment by negvet** [comment]
> I've talked with Evgeny, but in the tile detection case, it seems like many saliency maps are exported for each tile. So exception handle which `saliency_maps = [saliency_maps[i] for i in range(dataset.num_samples)]`, existed in the previous `develop`, can avoid the assertion error but maybe not make the appropriate saliency map(maybe saliency maps of first few tiles). Maybe saliency maps merging function or getting the saliency map from the original image(before tiling) would be needed.
> Yes, I'll update after the merge but I think we need some fixes in the saliency map for tiling case.. Maybe @negvet investigating and make another PR for it.

Actually it works just fine. Not that good, but functionally correct.
The first tile is the full image size - so `saliency_maps = [saliency_maps[i] for i in range(dataset.num_samples)]` actually provide the saliency maps for the whole images, which is correct (I deleted this line of code accidently, @dongkwan-kim01 correctly get it back).

The other problem here is that the model is not able to distinguish objects at full resolution (after being trained at tiles):
image
But it still can see the object occurrences (you can see the difference in object density at saliency maps) and can distinguish the background (it is SSD model)

It would be good to revisit this and merge saliency maps from tiles instead of generating it from full-res image. But for now it is fine IMO.

File: ote_sdk/ote_sdk/utils/tiler.py [hunk 2]
```diff
@@ -79,7 +80,7 @@ def predict(self, image: np.ndarray):
             detections = np.append(detections, output, axis=0)
             # cache full image feature vector and saliency map at 0 index
             if i == 0:
-                features = feats
+                features = copy.deepcopy(feats)
```
Comment (negvet) [inline ote_sdk/ote_sdk/utils/tiler.py:84]: `feats` is mutable, this helps to keep it from being overwritten when `i > 0`. This bug persisted before - seems that saliency maps (along with `feature_vector`) did not generate correctly before for the tiling case.

File: ote_sdk/ote_sdk/utils/tiler.py [hunk 3]
```diff
@@ -128,7 +129,7 @@ def predict_tile(
             if "feature_vector" in raw_predictions or "saliency_map" in raw_predictions:
                 features = [
                     raw_predictions["feature_vector"].reshape(-1),
-                    raw_predictions["saliency_map"],
+                    raw_predictions["saliency_map"][0],
                 ]
```
Comment (negvet) [inline ote_sdk/ote_sdk/utils/tiler.py:133]: Just removing the batch dimension, as in `BaseInferencerWithConverter.predict()`

**Review by wonjuleee — DISMISSED** [review]
LGTM

**Comment by dongkwan-kim01** [comment]
@negvet Thank you! So the original one was correct. Thank you for your confirmation.

**Review by harimkang — DISMISSED** [review]

**Review by sungmanc — DISMISSED** [review]

**Comment by dongkwan-kim01** [comment]
run ote-test

**Comment by dongkwan-kim01** [comment]
image
Since there some code change regarding Evgeny's commit and assertion revert, I ran the tile-detection in `tests/ote_cli/external/model-preparation-algorithm/ote_cli/test_tiling_detection.py` it locally and passed but also running the ote-test.

File: external/model-preparation-algorithm/tests/test_xai.py [hunk 4]
```diff
@@ -9,7 +9,6 @@
 from mmdet.models import build_detector
 
 from mpa.modules.hooks.auxiliary_hooks import ReciproCAMHook, DetSaliencyMapHook
-from mpa.det.stage import DetectionStage
```
Comment (negvet) [inline external/model-preparation-algorithm/tests/test_xai.py:13]: For me this test did not pass without this import locally. @dongkwan-kim01 could you please double check?

Comment (sungmanc) [reply external/model-preparation-algorithm/tests/test_xai.py:13]: @dongkwan-kim01 Could you check this? I want to merge this branch

Comment (dongkwan-kim01) [reply external/model-preparation-algorithm/tests/test_xai.py:13]: @negvet @sungmanc Yes I'll check it. Sorry for late response

Comment (dongkwan-kim01) [reply external/model-preparation-algorithm/tests/test_xai.py:13]: You're right it's making error. I've added with flake8 exception

**Review by sungmanc — DISMISSED** [review]

**Comment by sungmanc** [comment]
Test results: http://validationreports.sclab.intel.com:8006/reports/build_number_report?test_session_build_number=pr-ote-test-815&environment=iotg-dev-workstation-10

**Review by harimkang — DISMISSED** [review]

**Review by harimkang — APPROVED** [review]

**Review by sungmanc — APPROVED** [review]

**Pull request merged**